File: symmertric.py
"""101. Symmetric Tree
TC - O(n)
Sc - O(maxDepth)
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if root is None:
            return True
        return self.helper(root.left,root.right)
    
    def helper(self,t1,t2):
        if t1 == None and t2 == None:
            return True
        
        if t1 == None or t2 == None or t1.val != t2.val:
            return False
        
        return self.helper(t1.right,t2.left) and self.helper(t1.left,t2.right)
        
        
        

File: symmertric.java
//TC: O(n)
//SC: O(1)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean result;
    public boolean isSymmetric(TreeNode root) {
        result = true;
        helper(root.left, root.right);
        return result;
    }
    
    private void helper(TreeNode left, TreeNode right) {
        
        if(left == null && right == null) return;
        if(left == null || right == null) {
            result = false;
            return;
        }
        
        
        if(left.val != right.val) {
            result = false;
            return;
        }
        helper(left.left, right.right);
        helper(left.right, right.left);
    }
}